src/Microsoft.VisualStudio.LanguageServices.Razor/DefaultVisualStudioOpenDocumentManager.cs

```diff
@@ -0,0 +1,167 @@
+                    {
+                        // First time seeing this view import. Create a change tracker for it.
+                        viewImportChangeTracker = new ViewImportChangeTracker(
+                            _fileChangeService,
+                            _runningDocumentTable,
+                            _editorAdaptersFactoryService,
+                            _foregroundDispatcher,
+                            _errorReporter,
+                            importFilePath);
+
+                        _viewImportChangeTrackerCache[importFilePath] = viewImportChangeTracker;
+                    }
+                    else
+                    {
+                        viewImportChangeTracker = _viewImportChangeTrackerCache[importFilePath];
+                    }
+
+                    // We want the current document to be reparsed when this import is changed.
+                    viewImportChangeTracker.AddAssociatedParser(parser);
+                }
+            }
+        }
+
+        public override void RemoveDocument(VisualStudioDocumentTracker tracker)
+        {
+            _foregroundDispatcher.AssertForegroundThread();
+
+            if (_documents.Contains(tracker))
+            {
+                _documents.Remove(tracker);
+            }
+
+            if (!tracker.TextBuffer.Properties.TryGetProperty<VisualStudioRazorParser>(typeof(VisualStudioRazorParser), out var parser))
+            {
+                // The document tracker doesn't have a corresponding Razor parser. This should never be the case.
+                return;
+            }
+
+            EnsureUndisposedTrackers();
+
+            var currentDocumentFilePath = tracker.FilePath;
+            var imports = tracker.Imports;
+            foreach (var import in imports)
+            {
+                var importFilePath = import.FilePath;
+                if (importFilePath == null)
+                {
+                    // This is probably an in-memory view import. We can't track it.
+                    continue;
+                }
+
+                if (!_viewImportChangeTrackerCache.ContainsKey(importFilePath))
+                {
+                    // We were never tracking this file to begin with.
+                    continue;
+                }
+                var viewImportChangeTracker = _viewImportChangeTrackerCache[importFilePath];
+
+                // This document is being closed. It no longer needs to be reparsed.
+                viewImportChangeTracker.RemoveAssociatedParser(parser);
+            }
+        }
+
+        private void EnsureUndisposedTrackers()
+        {
+            foreach (var key in _viewImportChangeTrackerCache.Keys.ToList())
+            {
+                var value = _viewImportChangeTrackerCache[key];
+                if (value.IsDisposed)
+                {
+                    // This could happen if the file was deleted or renamed.
+                    _viewImportChangeTrackerCache.Remove(key);
```

Comment:
If that happens don't we get notified on another code path? If this is possible then I think we're missing something